Rename Readers to Decoders

Baccata opened the pull request:
For consistency across the codebase.

Baccata committed:
Rename Readers to Decoders

For consistency across the codebase.

daddykotex commented:
Love this effort

daddykotex commented on modules/aws-http4s/src/smithy4s/aws/internals/AwsEc2QueryCodecs.scala:
```diff
@@ -49,7 +49,7 @@ private[aws] object AwsEcsQueryCodecs {
 
   private val addErrorStartingPath = (_: Hints).add(XmlStartingPath(List("Response", "Errors", "Error")))
   private val discriminatorReaders =
```
`discriminatorDecoders`

I'm happy to make the changes as I do my review if you want, it will avoid comment spam

Baccata committed:
Remove inheritance relationship of capabilities

Adding inheritance relationship makes it harder to obtain non-orphaned
instances, and smithy4s does not aim at being a usable FP library. The
capabilities are only there for their abstraction value.

Baccata committed:
discriminatorReaders => discriminatorDecoders

Baccata commented on modules/core/src/smithy4s/capability/MonadThrowLike.scala:
```diff
@@ -18,7 +18,8 @@ package smithy4s.capability
 
 import smithy4s.kinds.PolyFunction
 
-trait MonadThrowLike[F[_]] extends Zipper[F] {
```
@daddykotex, this is a bit annoying for downstream, because it means that there are few more constructs to implement to use the generic routers, but it feels more "right", as otherwise there are some annoying bits wrt orphan instances, which make the decoders less usable.

Baccata replied:
Decided to move instances to orphan locations instead. It's more pragmatic.

daddykotex commented on modules/decline/src/core/OptsVisitor.scala:
```diff
@@ -129,7 +129,7 @@ object OptsVisitor extends SchemaVisitor[Opts] { self =>
 
   private def parseJson[A](schema: Schema[A]): String => Either[String, A] = {
-    val reader = smithy4s.json.Json.payloadCodecs.readers.fromSchema(schema)
+    val reader = smithy4s.json.Json.payloadCodecs.decoders.fromSchema(schema)
```
val decoder

Baccata committed:
Revert "Remove inheritance relationship of capabilities"

This reverts commit 6eea311e63b2b2b6992f0314093a7f56fd76533b.

Baccata committed:
Move capability instances to dedicated objects

This makes them orphan, but we do not want to have the responsibility
of making non-orphan instances for types we don't control, when the
capabilities form a subclass hierarchy.

Baccata committed:
headers

Baccata committed:
reader => decoder
